src/interface/mem.rs

```diff
@@ -0,0 +1,170 @@
+    let cage = cagetable_getref(cageid);
+
+    // only these four flags are allowed
+    let allowed_flags = (MAP_FIXED as i32 | MAP_SHARED as i32 | MAP_PRIVATE as i32 | MAP_ANONYMOUS as i32);
+    if flags & !allowed_flags > 0 {
+        // truncate flag to remove flags that are not allowed
+        flags &= allowed_flags;
+    }
+
+    if prot & PROT_EXEC > 0 {
+        return syscall_error(Errno::EINVAL, "mmap", "PROT_EXEC is not allowed");
+    }
+
+    // check if the provided address is multiple of pages
+    let rounded_addr = round_up_page(addr as u64);
+    if rounded_addr != addr as u64 {
+        return syscall_error(Errno::EINVAL, "mmap", "address it not aligned");
+    }
+
+    // offset should be non-negative and multiple of pages
+    if off < 0 {
+        return syscall_error(Errno::EINVAL, "mmap", "offset cannot be negative");
+    }
+    let rounded_off = round_up_page(off as u64);
+    if rounded_off != off as u64 {
+        return syscall_error(Errno::EINVAL, "mmap", "offset it not aligned");
+    }
+
+    // round up length to be multiple of pages
+    let rounded_length = round_up_page(len as u64);
+
+    let mut useraddr = addr as i32;
+    // if MAP_FIXED is not set, then we need to find an address for the user
+    if flags & MAP_FIXED as i32 == 0 {
+        let mut vmmap = cage.vmmap.write();
+        let result;
+        
+        // pick an address of appropriate size, anywhere
+        if addr as usize == 0 {
+            result = vmmap.find_map_space(rounded_length as u32 >> PAGESHIFT, 1);
+        } else {
+            // use address user provided as hint to find address
+            result = vmmap.find_map_space_with_hint(rounded_length as u32 >> PAGESHIFT, 1, addr as u32);
+        }
+
+        // did not find desired memory region
+        if result.is_none() {
+            return syscall_error(Errno::ENOMEM, "mmap", "no memory");
+        }
+
+        let space = result.unwrap();
+        useraddr = (space.start() << PAGESHIFT) as i32;
+    }
+
+    // TODO: validate useraddr (like checking whether within the program break)
+
+    flags |= MAP_FIXED as i32;
+
+    // either MAP_PRIVATE or MAP_SHARED should be set, but not both
+    if (flags & MAP_PRIVATE as i32 == 0) == (flags & MAP_SHARED as i32 == 0) {
+        return syscall_error(Errno::EINVAL, "mmap", "invalid flags");
+    }
+
+    let vmmap = cage.vmmap.read();
+
+    let sysaddr = vmmap.user_to_sys(useraddr);
+
+    drop(vmmap);
+
+    if rounded_length > 0 {
+        if flags & MAP_ANONYMOUS as i32 > 0 {
+            fildes = -1;
+        }
+
+        let result = cage.mmap_syscall(sysaddr as *mut u8, rounded_length as usize, prot, flags, fildes, off);
+        if result >= 0 {
+            if result != useraddr {
+                panic!("MAP_FIXED not fixed");
+            }
+
+            let mut vmmap = cage.vmmap.write();
+            let backing = {
+                MemoryBackingType::Anonymous
```

Comment:
whats going on here?

Reply:
This is marked TODO since shared memory file (shmget, etc) hasn't been integrated yet, so I am always using MemoryBackingType::Anonymous. It should be handled accordingly once the feature is implemented.

Reply:
file backed memory is an mmap function though, not shared memory?

Reply:
ok thats fine. Let's merge this into the main PR without doing that but mark that we need to. Chinmay can help with this more once we do that.